go/merge/three_way.go

```diff
@@ -200,7 +190,81 @@ func threeWayStructMerge(a, b, parent types.Struct, vwr types.ValueReadWriter, p
 		}
 		panic(fmt.Errorf("Bad key type in diff: %s", change.V.Type().Describe()))
 	}
-	return threeWayOrderedSequenceMerge(parent, aDiff, bDiff, makeGetFunc(a), makeGetFunc(b), makeGetFunc(parent), apply, vwr, progress)
+	return m.threeWayOrderedSequenceMerge(structCandidate{a}, structCandidate{b}, structCandidate{parent}, apply, path)
+}
+
+type candidate interface {
```

Comment:
> What is it that's distracting?

A few things, not necessarily in any order:
* PR diff is complicated
* figuring out whether these new collection types introduce any behavioral changes
* a fair amount of non-dense extra code
* boxing/unboxing is more code
* this entire algorithm is merging types.Collections and that's less obvious now
* I just find it less direct

I hear you on the type switching but I was more thinking like:

```go
differs := map[Kind]diffFunc{
  SetKind: func(v1, v2 types.Value, change chan<- types.ValueChanged, stop <-chan struct{}) types.Value {
    return v1.(Set).Diff(v2.(Set), change, stop)
  },
  MapKind: func(v1, v2 types.Value, change chan<- types.ValueChanged, stop <-chan struct{}) types.Value {
    return v1.(Map).Diff(v2.(Map), change, stop)
  },
  StructKind: func(v1, v2 types.Value, change chan<- types.ValueChanged, stop <-chan struct{}) types.Value {
    return v1.(Struct).Diff(v2.(Struct), change, stop)
  },
}
getters := map[Kind]getFunc{
  SetKind: func(v, k types.Value) types.Value {
    return k
  },
  MapKind: func(v, k types.Value) types.Value {
    return v.(Map).Get(k)
  },
  StructKind: func(v, k types.Value) types.Value {
    return v.(Struct).Get(string(k.(String)))
  },
}
pathConcat := map[Kind]pathConcatFunc { /* etc */ }
// then instead of aValue, bValue := a.get(aChange.V), b.get(bChange.V) have:
get := getters[a.Kind]
aValue, bValue := get(aChange.V), get(bChange.V)
```
now that I've typed that out I'm not sure it's any better. Is there a way to make the old callback approach more manageable? I can think of a few other ways to cut it.